src/model_theory/basic.lean

```diff
@@ -0,0 +1,340 @@
+/-
+Copyright (c) 2021 Aaron Anderson, Jesse Michael Han, Floris van Doorn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Aaron Anderson, Jesse Michael Han, Floris van Doorn
+-/
+import data.nat.basic
+
+/-!
+# Basics on First-Order Structures
+This file defines first-order languages and structures in the style of the
+[Flypitch project](https://flypitch.github.io/).
+
+## Main Definitions
+* A `first_order.language` defines a language as a pair of functions from the natural numbers to
+  `Type l`. One sends `n` to the type of `n`-ary functions, and the other sends `n` to the type of
+  `n`-ary relations.
+* A `first_order.language.Structure` interprets the symbols of a given `first_order.language` in the
+  context of a given type.
+* A `first_order.language.hom`, denoted `M →[L] N`, is a map from the `L`-structure `M` to the
+  `L`-structure `N` that commutes with the interpretations of functions, and which preserves the
+  interpretations of relations (although only in the forward direction).
+* A `first_order.language.embedding`, denoted `M ↪[L] N`, is an embedding from the `L`-structure `M`
+  to the `L`-structure `N` that commutes with the interpretations of functions, and which preserves
+  the interpretations of relations in both directions.
+* A `first_order.language.equiv`, denoted `M ≃[L] N`, is an equivalence from the `L`-structure `M`
+  to the `L`-structure `N` that commutes with the interpretations of functions, and which preserves
+  the interpretations of relations in both directions.
+
+## References
+For the Flypitch project:
+- [J. Han, F. van Doorn, *A formal proof of the independence of the continuum hypothesis*]
+[flypitch_cpp]
+- [J. Han, F. van Doorn, *A formalization of forcing and the unprovability of
+the continuum hypothesis*][flypitch_itp]
+
+-/
+
+universes l u
+
+variables {α : Type u} {β : Type u}
+
+namespace first_order
+
+/-- A first-order language consists of a type of functions of every natural-number arity and a
+  type of relations of every natural-number arity. -/
+structure language : Type (l+1) :=
+(functions : ℕ → Type l) (relations : ℕ → Type l)
+
+namespace language
+
+/-- The empty language has no symbols. -/
+def empty : language := ⟨λ _, pempty, λ _, pempty⟩
+
+instance : inhabited language := ⟨empty⟩
+
+/-- The type of constants in a given language. -/
+@[nolint has_inhabited_instance] def const (L : language.{l}) : Type l := L.functions 0
+
+variable (L : language.{u})
+
+/-- A language is relational when it has no function symbols. -/
+def is_relational : Prop := ∀ n, L.functions n → false
+
+/-- A language is algebraic when it has no relation symbols. -/
+def is_algebraic : Prop := ∀ n, L.relations n → false
```

Comment:
Have you considered making these classes?

Reply:
That's probably a good idea.